File: 6 семестр/ПиАПС/Lab 7/Lab 7/Models/CellState.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Lab_7.Models
{
    internal abstract class CellState : IObservable, IObserver
    {
        public List<Cell> Cells { get; init; }
        public bool IsMark { get; set; }
        public bool IsOpen { get; protected set; }

        public CellState()
        {
            Cells = new List<Cell>();
        }

        public void Attach(IObserver observer)
        {
            if (observer is Cell cell)
            {
                Cells.Add(cell);
            }
        }

        public void Detach(IObserver observer)
        {
            if (observer is Cell cell)
            {
                Cells.Remove(cell);
            }
        }

        public void Notify()
        {
            foreach (var cell in Cells)
            {
                if (GetBombCount() == 0)
                {
                    cell.Update(this);
                }
            }
        }

        public abstract void Update(IObservable subject);

        public void Open()
        {
            if (IsMark == false)
            {
                IsOpen = true;
            }
        }

        public bool RightClick()
        {
            if (IsOpen == false)
            {
                IsMark = !IsMark;
            }
            return IsMark;
        }

        public abstract int Click();
        public abstract int GetBombCount();
    }
}

File: 11 semester/Games/LR_5/Source/LR_5/CellState.cpp
// Fill out your copyright notice in the Description page of Project Settings.


#include "CellState.h"

CellState::~CellState()
{
}

CellState::CellState() : IsMark(false), IsOpen(false) {}

void CellState::Detach(IObserver* observer)
{
    Cell* cell = dynamic_cast<Cell*>(observer);
    if (cell) {
        Cells.erase(std::remove(Cells.begin(), Cells.end(), cell), Cells.end());
    }
}

void CellState::Notify()
{
    for (auto cell : Cells) {
        if (GetBombCount() == 0) {
            cell->Update(this);
        }
    }
}

void CellState::Open() {
    if (!IsMark) {
        IsOpen = true;
    }
}

bool CellState::RightClick() {
    if (!IsOpen) {
        IsMark = !IsMark;
    }
    return IsMark;
}

void CellState::Attach(IObserver* observer)
{
    Cell* cell = dynamic_cast<Cell*>(observer);
    if (cell) {
        Cells.push_back(cell);
    }
}
